src/irmin-pack/IO.ml

```diff
@@ -146,10 +185,32 @@ module Unix : S = struct
     in
     aux dirname (fun () -> ())
 
+  let raw ~mode ~version file =
+    let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
+    let raw = Raw.v x in
+    let header = { Raw.Header.version; offset = 0L; generation = 0L } in
+    Raw.Header.set raw header;
+    raw
+
   let clear t =
-    t.offset <- 0L;
-    t.flushed <- header;
-    Buffer.clear t.buf
+    Log.debug (fun l -> l "clear %s" t.file);
+    if t.readonly then invalid_arg "Read-only IO cannot be cleared";
+    Buffer.clear t.buf;
+    (* no-op if the file is already empty; this is to avoid bumping
+       the version number when this is not necessary. *)
+    if t.offset = 0L then ()
+    else (
+      t.offset <- 0L;
+      t.flushed <- header t;
+      if t.version = `V1 then t.version <- `V2;
+      t.generation <- Int64.succ t.generation;
+      (* set the header, because calling flush after clear will not set it. *)
+      Raw.Header.set t.raw
+        {
+          Raw.Header.version = bin_of_version t.version;
+          offset = t.offset;
+          generation = t.generation;
+        })
```

Comment:
Note that `flush` also races with `clear` now, since `flush` doesn't get header information in the same order as `clear`. (I'm not sure if that's observable from `pack` – perhaps `flush` also wants the lock.)

Reply:
Would be nice if we could find a case where this race happens. I'll try to come up with a few tests.